postgres performance improvements - Write path

dannyzaken opened the pull request:
This PR improves most of the queries in the write path. Read path is not handled in this PR

### Explain the changes

#### in postgres_client
* Added option to enable query logging using env `PG_ENABLE_QUERY_LOG`. this will print for each query
  the time it took and if enabled (with `PG_EXPLAIN_QUERIES`) the EXPLAIN result of the query
* fixes in encode_json and handle_ops_encoding
  * encode mongo.objectid as string in postgres_client encode_json - when mongo object id is passed as object to mongo-to-postgres lib, it will convert the query to access data->_id as a jsonb struct. this causes queries to not hit the indexes. when passed as string the lib generates the query using `->>` that better matches the index
  * handle_ops_encoding - handle multiple ops in one query
  * handle comparison operators (`$eq`, `$ne`, etc.)
* add `{disableContainmentQuery: true}` when calling `mongo-query-to-postgres-jsonb`
* fix sorts to use indexes
  * access `_id` as text
  * remove NULLS FIRST\LAST and use default as we do in the indexes

#### other changes
* added datachunk index by _id descending for iterate_all_chunks
* in md_store - pass `$exists` in some queries to match indexes
* in system_server - cache result of list_undeletable_buckets
  to avoid calling it on each read_system
* bumped version of `mongo-query-to-postgres-jsonb` to `0.2.14` 


### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1.

dannyzaken committed:
postgresdb performance fixes

* Added option to enable query logging. this will print for each query
  the time it took and if enabled the EXPLAIN result of the query
* fixes in encode_json and handle_ops_encoding
  * encode mongo id as string to match indexes
  * handle multiple ops in one query
  * handle comparison ops
* in md_store - pass $exists in some queries to match indexes
* in system_server - cache result of list_undeletable_buckets
  to avoid calling it on each read system

* postgres_client - fix sorts to use indexes
  * access _id as text
  * remove NULLS FIRST\LAST and use default as we do in the indexes
* added datachunk index by _id descending for iterate_all_chunks
* added $exists in MDStore.load_parts_objects_for_chunks

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

encode mongo.objectid as string in postgres_client encode_json

when mongo object id is passed as object to imongo-to-postgres lib, it will convert the query to access data->_id as a jsonb struct. this causes queries to not hit the indexes

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

added {$exists: true} to chunks query in load_blocks_for_chunks

this will cause the postgres query to match the filter in the index

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

added $exists for dedup_keys in find_chunks_by_dedup_key

* without the $exists the query does not hit the index

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

modified handle_ops_encoding to handle multiple ops

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

fixed find_all_parts_of_object to hit index

* added {$exists: true } to obj
* fixes handle_ops_encoding for comparison operators ($eq etc.)

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

cache results of bucket_server.list_undeletable_buckets() in read_system

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

db query log changes

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

Update postgres_client.js

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

use {disableContainmentQuery: true} option when calling mongo_to_pg

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

more postgres improvements

* postgres_client - fix sorts to use indexes
  * access _id as text
  * remove NULLS FIRST\LAST and use default as we do in the indexes
* added datachunk index by _id descending for iterate_all_chunks
* added $exists in MDStore.load_parts_objects_for_chunks

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

mark some indexes as postgres only and skip in mongo_client

Update postgres_client.js

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

Update md_store.js

guymguym commented on src/util/postgres_client.js:
```diff
@@ -411,7 +448,16 @@ class PostgresTable {
         this.name = name;
         this.init_function = init_function;
-        this.db_indexes = db_indexes;
+        this.db_indexes = _.concat([{
+                fields: {
+                    _id: 1,
+                },
+                options: {
+                    name: '_id_index',
+                    unique: true,
+                }
+            }],
+            db_indexes || []);
```
I would replace concat with `[item, ...arr]`
```suggestion
        const id_index = {
            fields: {
                _id: 1,
            },
            options: {
                name: '_id_index',
                unique: true,
            }
        };
        this.db_indexes = [id_index,  ...(db_indexes || [])];
```

guymguym commented on src/util/postgres_client.js:
```diff
@@ -149,13 +153,46 @@ function handle_ops_encoding(schema, val) {
+        try {
+            await log_query(pg_client, q, tag, millitook, pool_count, transaction_counter === 0);
+        } catch (err) {
+            console.error('error on log query', err);
```
can log_query fail?

dannyzaken replied:
it can fail when running explain

guymguym replied:
ah I thought I saw you had try-catch inside

guymguym commented on src/util/postgres_client.js:
```diff
@@ -149,13 +153,46 @@ function handle_ops_encoding(schema, val) {
-        dbg.log1(`postgres_client: ${tag}: got result:`, util.inspect(res, { depth: 6 }));
+        const millitook = time_utils.millitook(millistart);
+        try {
+            await log_query(pg_client, q, tag, millitook, pool_count, transaction_counter === 0);
```
Can you add a comment to explain why `should_explain = transaction_counter === 0` ?
It's not clear to me.

dannyzaken replied:
sure. 
In general the log_query code is a quick and dirty solution. I plan to refactor it if and before merging. the if is there since failures in the explain for transaction caused the transaction to no longer be valid, so I decided for now to just avoid it

guymguym replied:
got it. I think it's worth keeping instrumentation inside the code behind env var, so we can use for next performance improves. or you have a better way to collect those from `pg_stat_statements`?
perhaps keep such scripts in git too?

guymguym commented on src/util/postgres_client.js:
```diff
@@ -149,13 +153,46 @@ function handle_ops_encoding(schema, val) {
+    const explain_q = { text: 'explain ' + q.text, values: q.values };
+    let explain_rows = ['explain:\tno explain'];
+    try {
+        if (should_explain && process.env.EXPLAIN_QUERIES === 'true') {
```
I would rename the env to PG_EXPLAIN_QUERIES or DB_EXPLAIN_QUERIES

guymguym commented on src/util/postgres_client.js:
```diff
@@ -149,13 +153,46 @@ function handle_ops_encoding(schema, val) {
+        explain_rows = ['explain:\texplain failed'];
+    }
+
+    let stack = (new Error()).stack.split('\n').slice(3).join(',');
+    const msgs = [];
+    msgs.push(`stack: ${stack}`);
+    msgs.push(`tag: ${tag}`);
+    msgs.push(`query: ${q.text}`);
+    msgs.push(`took: ${millitook}`);
+    msgs.push(`pool: ${pool_count}`);
+    msgs.push(`explain:`);
+    let log_message = msgs.concat(explain_rows).join('\nQUERY_LOG:\t');
+    console.log(`\nQUERY_LOG:\t${log_message}`);
```
There seems to be q lot of objects being created here just for logging.
I would try to keep this more straightforward code that prints the string.
Also I think it would be best to keep the format as a json object and not a multiline log print.

guymguym reviewed (approved):
small comments, otherwise looks good

dannyzaken commented:
try to ignore the first commit. It will not remain as is

dannyzaken committed:
bumped mongo-query-to-postgres-jsonb to 0.2.14

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>

dannyzaken committed:
added option to run s3perf --get without specific key

in key is omitted the bucket is listed and the object list is iterated with getObject

Signed-off-by: Danny Zaken <dannyzaken@gmail.com>
